### PR TITLE
JDK-8299672: Enhance HeapDump JFR event

### PR DESCRIPTION
**Pull request by MBaesken**
Enhance the JFR Event HeapDump  with the additional interesting fields, compression and overwrite.
Add some UL logging in case the heap dump writing failed .

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8299672](https://bugs.openjdk.org/browse/JDK-8299672): Enhance HeapDump JFR event

### Reviewers
 * [Ralf Schmelter](https://openjdk.org/census#rschmelter) (@schmelter-sap - **Reviewer**) ⚠️ Review applies to [750ccf90](https://git.openjdk.org/jdk/pull/11864/files/750ccf90d744541fee6d83f0bdcc81a5c1b23869)
 * [Christoph Langer](https://openjdk.org/census#clanger) (@RealCLanger - **Reviewer**) ⚠️ Review applies to [76fc3a9c](https://git.openjdk.org/jdk/pull/11864/files/76fc3a9c236c508f987a7b40d6a006a32817e46e)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11864/head:pull/11864` \
`$ git checkout pull/11864`

Update a local copy of the PR: \
`$ git checkout pull/11864` \
`$ git pull https://git.openjdk.org/jdk pull/11864/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11864`

View PR using the GUI difftool: \
`$ git pr show -t 11864`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11864.diff">https://git.openjdk.org/jdk/pull/11864.diff</a>

</details>

### CODE REVIEW

**Review by schmelter-sap — APPROVED** [review]
Looks good.

**Comment by egahlin** [comment]
What's the rationale for adding compression level and overwrite?

**Comment by MBaesken** [comment]
The support for gzipped heap dumps was added recently with "8237354: Add option to jcmd to write a gzipped heap dump" but not reflected in JFR so I wanted to add it here too to get a full picture of the heap dump creation.
The overwrite field was just added  for completion, no specific reason.

File: test/jdk/jdk/jfr/event/diagnostics/TestHeapDump.java [hunk 1]
```diff
@@ -65,6 +65,8 @@ public static void main(String[] args) throws Exception {
             Events.assertField(e, "gcBeforeDump").equal(true);
             Events.assertField(e, "onOutOfMemoryError").equal(false);
             Events.assertField(e, "size").equal(Files.size(path));
+            Events.assertField(e, "compression").equal(-1);
```
[the remark this answers is not in the document]
Comment (MBaesken) [reply test/jdk/jdk/jfr/event/diagnostics/TestHeapDump.java]: Hi Ralf, thanks for the review. 
I adjusted the test as suggested.

File: src/hotspot/share/jfr/metadata/metadata.xml [hunk 2]
```diff
@@ -1144,11 +1144,14 @@
     <Field type="ulong" contentType="bytes" name="size" label="Size Written" />
   </Event>
 
-  <Event name="HeapDump" category="Java Virtual Machine, Diagnostics" label="Heap Dump" stackTrace="true" thread="true">
-    <Field type="string" name="destination" label="Destination" />
+  <Event name="HeapDump" category="Java Virtual Machine, Diagnostics" label="Heap Dump" description="Information about a successfully written Java Heap dump"
```
Comment (RealCLanger) [inline src/hotspot/share/jfr/metadata/metadata.xml:1147]: spelling: "Java heap dump" in description (small heap)

File: src/hotspot/share/jfr/metadata/metadata.xml [hunk 3]
```diff
@@ -1144,11 +1144,14 @@
...
+    stackTrace="true" thread="true">
+    <Field type="string" name="destination" label="Destination path of the dump" />
```
Comment (RealCLanger) [inline src/hotspot/share/jfr/metadata/metadata.xml:1149]: Spelling: In the label we'd write Path and Dump in capitals

File: src/hotspot/share/jfr/metadata/metadata.xml [hunk 4]
```diff
@@ -1144,11 +1144,14 @@
...
     <Field type="long" name="size" label="Size" />
     <Field type="boolean" name="gcBeforeDump" label="GC Before Dump" />
-    <Field type="boolean" name="onOutOfMemoryError" label="On Out of Memory Error" />
+    <Field type="boolean" name="onOutOfMemoryError" label="Heap dump on Out of Memory Error" />
```
Comment (RealCLanger) [inline src/hotspot/share/jfr/metadata/metadata.xml:1152]: Also here: "Dump" in capitals in the label

File: src/hotspot/share/jfr/metadata/metadata.xml [hunk 5]
```diff
@@ -1144,11 +1144,14 @@
...
+    <Field type="boolean" name="overwrite" label="Overwrite" description="Heap dump overwrites previous file location if it exists" />
+    <Field type="int" name="compression" label="Compression Level" description="Compression level of the Dump, if larger than 0 we use gzip compressoion with this level" />
```
Comment (RealCLanger) [inline src/hotspot/share/jfr/metadata/metadata.xml:1154]: Here, in description, use small letter for "dump"

**Review by RealCLanger — CHANGES_REQUESTED** [review]
Some spelling issues, for consistency.

**Comment by MBaesken** [comment]
Hi Christoph, I did some adjustments (also compressoion  -> compression).

**Review by RealCLanger — APPROVED** [review]
Looks good to me now.

PS: The metadata.xml also has a copyright year to update which you should do before integrating.

**Comment by egahlin** [comment]
> The support for gzipped heap dumps was added recently with "8237354: Add option to jcmd to write a gzipped heap dump" but not reflected in JFR so I wanted to add it here too to get a full picture of the heap dump creation.
> The overwrite field was just added for completion, no specific reason.

Not sure what the user is going to do with the information, but OK.

**Pull request closed without merging**